### PR TITLE
feat: supporting multiple consensus instances

### PR DESCRIPTION
**Pull request by b00f**
## Description

User will be able to run multiple consensus instances (validator) in a single node

### CODE REVIEW

File: node/config/example_config.toml [hunk 1]
```diff
@@ -1,138 +1,132 @@
...
+# `store` contains configuration options for the store module, which manages storage and retrieval of blockchain data.
 [store]
 
-  # Path specifies the data directory.
+  # `path` specifies the directory where blockchain data will be stored.
   # Default is data
  ## path = "data"
 
-# network contains configuration of network module.
+# `network` contains configuration options for the network module, # which manages communication between nodes.
```
Comment (themantre) [inline node/config/example_config.toml:13]: # which => which

File: node/config/example_config.toml [hunk 2]
```diff
@@ -1,138 +1,132 @@
...
+    # `network.bootstrap` contains configuration for bootstrapping the node.
   [network.bootstrap]
 
-    # addresses is the list of peers addresses needed for peer discovery.
+    # `addresses` is a list of peer addresses needed for peer discovery.
+    # These addresses are used by the Pactus node to discover and connect to other peers on the network.
    ## addresses = ["/ip4/172.104.46.145/tcp/21777/p2p/12D3KooWNYD4bB82YZRXv6oNyYPwc5ozabx2epv75ATV3D8VD3Mq"]
 
-    # min_threshold is the number of connections it attempts to maintain.
+    # `min_threshold` is the minimum number of connections that the Pactus node can maintain.
```
Comment (themantre) [inline node/config/example_config.toml:55]: can maintain => should maintain

File: state/state.go [hunk 3]
```diff
@@ -459,35 +441,37 @@ func (st *state) CommitBlock(height uint32, block *block.Block, cert *block.Cert
...
+		ok, proof := sortition.EvaluateSortition(st.lastInfo.SortitionSeed(), signer, st.totalPower(), val.Power())
+		if ok {
+			trx := tx.NewSortitionTx(st.lastInfo.BlockHash().Stamp(), val.Sequence()+1, val.Address(), proof)
+			signer.SignMsg(trx)
+
+			err := st.txPool.AppendTxAndBroadcast(trx)
+			if err == nil {
+				st.logger.Info("sortition transaction broadcasted",
+					"address", signer.Address(), "power", val.Power(), "tx", trx)
+				return true
```
Comment (themantre) [inline state/state.go:470]: should continue

File: state/state_test.go [hunk 4]
```diff
@@ -386,39 +375,39 @@ func TestSortition(t *testing.T) {
...
 	assert.True(t, tState1.committee.Contains(tValSigner1.Address()))
 	assert.True(t, tState1.committee.Contains(pub.Address()))
 
 	// ---------------------------------------------
 	// Let's save and load tState1
 	tState1.Close()
-	state1, _ := LoadOrNewState(tState1.config, tState1.genDoc, tValSigner1, store, tCommonTxPool, nil)
-	st2 := state1.(*state)
+	st, _ = LoadOrNewState(tState1.genDoc, []crypto.Signer{tValSigner1}, store, tCommonTxPool, nil)
+	stVal1 := st.(*state)
```
Comment (themantre) [inline state/state_test.go:406]: rename `stVal1` to `st1` and `stNewVal` to `stNew`

File: state/state_test.go [hunk 5]
```diff
@@ -676,3 +671,19 @@ func TestCalculatingGenesisState(t *testing.T) {
...
+
+	moveToNextHeightForAllStates(t)
+
+	txs := block.NewTxs()
+	trx := tState2.createSubsidyTx(crypto.GenerateTestAddress(), 0)
+	txs.Append(trx)
+	b := block.MakeBlock(2, util.Now(), txs, tState2.lastInfo.BlockHash(), tState2.stateRoot(), tState2.lastInfo.Certificate(), tState2.lastInfo.SortitionSeed(), tState2.signers[0].Address())
+	c := makeCertificateAndSign(t, b.Hash(), 0, tValSigner1, tValSigner2, tValSigner3, tValSigner4)
+
+	// If tState1 receives a block with version 2 and rejects it,
```
Comment (themantre) [inline state/state_test.go:686]: // tState1 receives a block with version 2 and rejects it.
	// It is possible that the same block would be considered valid by tState2.

File: sync/sync.go [hunk 6]
```diff
@@ -409,13 +400,19 @@ func (sync *synchronizer) peerIsInTheCommittee(id peer.ID) bool {
 		return false
 	}
 
-	return sync.state.IsInCommittee(p.Address())
+	for key := range p.ConsensusKeys {
```
Comment (themantre) [inline sync/sync.go:403]: Consensus also can tell us if we have an active validator

**Review by themantre — CHANGES_REQUESTED** [review]

File: consensus/mock.go [hunk 7]
```diff
@@ -12,19 +10,36 @@ import (
...
-	State    *state.MockState
+	Active   bool
 	Height   uint32
 	Round    int16
 }
 
-func MockingConsensus(state *state.MockState) *MockConsensus {
-	return &MockConsensus{State: state}
+func MockingManager(signers []crypto.Signer) (Manager, []*MockConsensus) {
+	mocks := make([]*MockConsensus, len(signers))
```
[the remark this answers is not in the document]
Comment (b00f) [reply consensus/mock.go:22]: It is useful for tests

File: sync/sync.go (hunk 6 above)
Comment (b00f) [reply sync/sync.go:403]: No, this is about other peers in the network.

**Review by themantre — APPROVED** [review]

**Pull request merged**